### DISCUSSION TITLE
CRUD Operations, Templates & Third-Party Components

### DISCUSSION THREAD

**Discussion by Sharafudeen** [oqtane/oqtane.framework#1116] [General] [answered]
Hi,

This is great framework for developers who is developing Modern Web Application.
I have few questions to start with Oqtane Framework,

1. How do I create CRUD gridview?
2. How do I create Index, Create, Edit, Details & Delete pages? (Ex., Employee entity)
3. How do I enable and disabled Multitenant for my whole application?
4. Can I integrate Syncfusion Blazor components?
5. Can I implement completely New Admin or Frontend Templates?

Thanks & Regards!

**Comment by hishamco** [comment]
> How do I create CRUD gridview?

You can integrate with any open source gridview, I remembered we already implemented a GridView in [BlazorWebFormsComponents](https://github.com/FritzAndFriends/BlazorWebFormsComponents) but you are free to choose what you like

> How do I create Index, Create, Edit, Details & Delete pages? (Ex., Employee entity)

Have a look to [Blog](https://github.com/oqtane/oqtane.blogs) module

> How do I enable and disabled Multitenant for my whole application?

Oqtane is built with multitenance feature, I'm not sure if you can disabled, but I think no, @sbwalker correct me if I'm wrong

> Can I integrate Syncfusion Blazor components?

@ADefWebserver is already mentioned that somewhere, also I think he blogged about it

> Can I implement completely New Admin or Frontend Templates?

Sure, there's already two themes with the default installation

**Comment by sbwalker** [comment, marked as answer]
1. How do I create CRUD gridview?

>>  I assume you are referring to a component which displays a listing of items which allows you to add, edit, delete individual items. All of the current admin modules in the framework contain an Index.razor component which contains this type of functionality ( ie. https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Client/Modules/Admin/Roles/Index.razor ). If you want to create a custom module I would suggest you use the Module Creator as it scaffolds a functional CRUD module for you. You can learn about the Module Creator at the 19 minute point of this training video: https://www.youtube.com/watch?v=JPfUZPlRRCE&list=PLYhXmd7yV0elLNLfQwZBUlM7ZSMYPTZ_f&index=1

2. How do I create Index, Create, Edit, Details & Delete pages? (Ex., Employee entity)

>> You don't create pages in Oqtane - you create razor components. If you use the Module Creator mentioned above it will provide add, edit, delete functionality for you.

3. How do I enable and disabled Multitenant for my whole application?

>> There is no option to enable or disable multi-tenancy, it is part of the framework. If you only want to manage a single tenant in your installation then simply use the default Site created during installation.

4. Can I integrate Syncfusion Blazor components?

>> Absolutely. You can learn more on this blog: https://blazorhelpwebsite.com/ViewBlogPost/47

5. Can I implement completely New Admin or Frontend Templates?

>> Yes. You can see a custom template project here: https://github.com/oqtane/oqtane.theme.cerulean

**Reply by Sharafudeen** [reply]
Module Creator creates Razor Components for CRUD Operations. How do I do if I have additional features such as Approve, Reject, etc., for the Custom Module?

1. Do I need to create new Razor Component for Approve.razor in the custom module project?
2. If I create as above, then the approve permission will appear in the Oqtane Admin Control Panel?

Thanks!

**Reply by sbwalker** [reply]
I am not sure if I understand your question. If you are asking how you could extend a custom module created by the Module Creator to allow for Approve/Reject of individual items, then you would need to modify the relevant code. At a minimum you would most likely need to add an IsApproved field to the database table and a property to the model. For the user interface there are many ways to approach this depending on your preference and needs. For example you could simply add a checkbox to the Edit.razor so a user can choose if an item is approved or not.

I have no idea what you are referring to about "permissions in the Oqtane Admin Control Panel".

**Reply by Sharafudeen** [reply]
Thanks for your detailed information.

"Permissions in the Oqtane Admin Control Panel" - Every modules contains permissions for CRUD operations under "Manage Settings" in the Oqtane "Admin Dashobard". Please refer attached screenshot. But it contains only Edit & View permissions for the listed roles? 

1. How can I add more fields as a Column (for approve/reject)?
2. I don't need it for all modules, I need it for custom modules only.
3. This feature can apply across sites and tenants.
4. What do you suggest for this feature? Is it really need to add as a column or we can manage in the custom module level?

_**Note:** This feature requires for CMS website and other workflow applications_


![settings-permission](https://user-images.githubusercontent.com/42769344/108310915-37ecb100-71da-11eb-9ad3-e9d204a01e2a.PNG)

Thanks!

**Reply by sbwalker** [reply]
If you need to add a custom permission for your module, you can specify a value for the PermissionNames property in the IModule interface:

```
    public class ModuleInfo : IModule
    {
        public ModuleDefinition ModuleDefinition => new ModuleDefinition
        {
            Name = "Custom Module",
            Description = "My Custom Module",
            Version = "1.0.0",
            PermissionNames = Shared.PermissionNames.View + "," + Shared.PermissionNames.Edit + ",Approve" 
        };
    }
```

This will add an "Approve" column to the permission grid in the Module Settings UI ( in addition to View and Edit ) and will allow a user to make their own selections. The permission grid will take care of saving these selections automatically.

Then in your client razor component code you will need to use the following method to check if the current user is authorized:

```
UserSecurity.IsAuthorized(PageState.User,"Approve", ModuleState.Permissions)
```

And in your server controllers you can use

```
UserPermissions.IsAuthorized(User, EntityNames.Module, moduleid, "Approve")
```

I hope this helps.

**Reply by Sharafudeen** [reply]
This looks custom feature extensible is very easy in Oqtane Framework. Thanks for sharing the code!

**Comment by Sharafudeen** [comment]
Thanks for you quickest response.
Oqtane is simply an awesome framework and surely it becomes an ecosystem!
